pkg/ingester/ingester.go

```diff
@@ -247,7 +254,38 @@ func (i *Ingester) Query(req *logproto.QueryRequest, queryServer logproto.Querie
 	}
 
 	instance := i.getOrCreateInstance(instanceID)
-	return instance.Query(req, queryServer)
+	err = instance.Query(req, queryServer)
+	if err != nil {
+		return err
+	}
+
+	if i.cfg.QueryStore {
+		start := req.Start
+		end := req.End
+		if i.cfg.QueryStoreMaxLookBackPeriod != 0 {
+			oldestStartTime := time.Now().Add(-i.cfg.QueryStoreMaxLookBackPeriod)
+			if oldestStartTime.After(req.Start) {
+				start = oldestStartTime
+			}
+		}
+
+		if start.Before(end) {
+			req.Start = start
+			req.End = end
+
+			itr, err := i.store.LazyQuery(queryServer.Context(), logql.SelectParams{QueryRequest: req})
+			if err != nil {
+				return err
+			}
+
+			err = sendBatches(queryServer.Context(), itr, queryServer, req.Limit)
```

Comment:
I have some concerns here.

1 - You're sending data directly and stats won't reflect that change, you need some refactoring to support sending with stats.
2 - Even if you do support stats, I'm not sure you can send the data from the store and the ingester via GRPC sequentially. I could be wrong but I think you're losing ordering by not considering and deduping line prior sending them. In over words there could be overlap of data.

I'm not 100% but you'll need to build tests for this. if you keep it that way.

Reply:
I was under the impression that whatever duplicates would be there querier would dedup it. I will investigate more. I feel here order of entries is more important.

Reply:
Querier dedupe per batches:

```golang
// NewQueryClientIterator returns an iterator over a QueryClient.
func NewQueryClientIterator(client logproto.Querier_QueryClient, direction logproto.Direction) EntryIterator {
	return &queryClientIterator{
		client:    client,
		direction: direction,
	}
}

func (i *queryClientIterator) Next() bool {
	for i.curr == nil || !i.curr.Next() {
		batch, err := i.client.Recv()
		if err == io.EOF {
			return false
		} else if err != nil {
			i.err = err
			return false
		}

		i.curr = NewQueryResponseIterator(i.client.Context(), batch, i.direction)
	}

	return true
}
```

Imagine the second batch contains half of the data of the first batch. You'll end up not deduping this data. You can use that iterator in a test to prove it's fine.